[EN, ZH] Improve `Beatmap/HP drain rate` and add `Beatmap/HP drain rate (lazer)` (#14936)
ppy/osu-wiki: wiki/Beatmap/HP_drain_rate_(lazer)/en.md

---
stub: true
tags:
  - health drain rate
  - passive drain rate
---

> # HP drain rate

Reviewer: ```suggestion
# HP drain rate (lazer)
```

> # HP drain rate

Reviewer 2: ```suggestion
# HP drain rate (lazer)
```

Author: In the game, this attribute name has been changed to "HP Drain". Should I also update the page (e.g., the title, internal links, or terminology) accordingly?
<img width="1768" height="552" alt="image" src="https://github.com/user-attachments/assets/b768b19e-85ee-43fc-a6cf-b5aabc7232a3" />

> *For the osu!stable version of this article, see: [HP drain rate](/wiki/Beatmap/HP_drain_rate)*\

Reviewer 2: ```suggestion
*For the osu!(stable) version of this article, see: [HP drain rate](/wiki/Beatmap/HP_drain_rate)*\
```

should also link this lazer article in the stable version

*For the gameplay meaning of health, see: [Health](/wiki/Gameplay/Health)*\

> *For the ranking criteria related to HP drain, see: [Ranking criteria](/wiki/Ranking_criteria)*

Reviewer 2: since there's no specific rc section on hp drain, this hatnote doesn't make any sense

Author: Does the stable part also need to be changed

Author: > since there's no specific rc section on hp drain, this hatnote doesn't make any sense

<img width="724" height="553" alt="image" src="https://github.com/user-attachments/assets/369f6854-ef13-4dc5-9a84-9757c1bfd1dc" />

In some modes (osu!, osu!taiko, osu!catch), RC still has requirements for HP values, while only osu!mania has no requirements for HP values

**HP drain rate** (***HP***) is a [beatmap](/wiki/Beatmap) difficulty setting that controls how much [health](/wiki/Gameplay/Health) is passively lost to health drain. It also affects how heavily a player is penalised for missing notes and how much health is gained back by accurately hitting [hit objects](/wiki/Gameplay/Hit_object).

HP ranges from 0 to 10, (extendable to 11 via the [Difficulty Adjust (DA)](/wiki/Gameplay/Game_modifier/Difficulty_Adjust) mod), where 10 (or 11) is the most punishing and 0 is the most lenient.

## Performance in various game modes

### osu!

<!-- TODO: quantify the relationship between passive HP drain rate and the value -->

> In [osu!](/wiki/Game_mode/osu!), the HP value affects both the passive drain rate and the judgement penalties/rewards.\

Reviewer 2: leave blank lines between paragraphs, don't join them with backslashes

| **PERFECT** | Any | `(0.0055 - HP * 0.0005) * 100%` | +0.55% | +0.30% | +0.05% | **0.00%** |
| **GREAT** | Any | `(0.0050 - HP * 0.0005) * 100%` | +0.50% | +0.25% | **0.00%** | -0.05% |
| **GOOD** | Any | `(0.0040 - HP * 0.0004) * 100%` | +0.40% | +0.20% | **0.00%** | -0.04% |
| **OK** | Any | `0` | 0.00% | 0.00% | 0.00% | 0.00% |
| **MEH** | Any | `[-(HP + 1) * 0.0016] * 100%` | -0.16% | -0.96% | -1.76% | -1.92% |
| **MISS** | Normal note | `[-(HP + 1) * 0.0075] * 100%` | -0.75% | -4.50% | -8.25% | -9.00% |
| **MISS** | Head or tail of a [hold note](/wiki/Beatmap/Pattern/osu!mania/Hold_note) | `[-(HP + 1) * 0.00375] * 100%` | -0.375% | -2.25% | -4.125% | -4.50% |

## Mod effects

There are seven mods that alter the HP drain rate when activated:

- [Easy (EZ)](/wiki/Gameplay/Game_modifier/Easy_(lazer)): Halves the HP value.
- [Hard Rock (HR)](/wiki/Gameplay/Game_modifier/Hard_Rock_(lazer)): Multiplies the HP value by 1.4, up to a maximum of 10.
- [Double Time (DT)](/wiki/Gameplay/Game_modifier/Double_Time_(lazer))/[Nightcore (NC)](/wiki/Gameplay/Game_modifier/Nightcore_(lazer)): The HP value is not affected, but due to the 50% play speed increase, health drain occurs 50% faster.

> - [Half Time (HT)](/wiki/Gameplay/Game_modifier/Half_Time_(lazer))/[Daycore (DC)](/wiki/Gameplay/Game_modifier/Daycore): The HP value is not affected, but due to the 25% play speed decrease, health drain occurs 25% slower.

Reviewer 3: ```suggestion
- [Double Time (DT)](/wiki/Gameplay/Game_modifier/Double_Time_(lazer)) / [Nightcore (NC)](/wiki/Gameplay/Game_modifier/Nightcore_(lazer)): The HP value is not affected, but due to the 50% play speed increase, health drain occurs 50% faster.
- [Half Time (HT)](/wiki/Gameplay/Game_modifier/Half_Time_(lazer)) / [Daycore (DC)](/wiki/Gameplay/Game_modifier/Daycore): The HP value is not affected, but due to the 25% play speed decrease, health drain occurs 25% slower.
```
ditto

> - [Half Time (HT)](/wiki/Gameplay/Game_modifier/Half_Time)/[Daycore (DC)](/wiki/Gameplay/Game_modifier/Daycore): The HP value is not affected, but due to the 25% play speed decrease, health drain occurs 25% slower.

Reviewer: Since this is a lazer article now
```suggestion
- [Easy (EZ)](/wiki/Gameplay/Game_modifier/Easy_(lazer)): Halves the HP value.
- [Hard Rock (HR)](/wiki/Gameplay/Game_modifier/Hard_Rock_(lazer)): Multiplies the HP value by 1.4, up to a maximum of 10.
- [Double Time (DT)](/wiki/Gameplay/Game_modifier/Double_Time_(lazer))/[Nightcore (NC)](/wiki/Gameplay/Game_modifier/Nightcore_(lazer)): The HP value is not affected, but due to the 50% play speed increase, health drain occurs 50% faster.
- [Half Time (HT)](/wiki/Gameplay/Game_modifier/Half_Time_(lazer))/[Daycore (DC)](/wiki/Gameplay/Game_modifier/Daycore): The HP value is not affected, but due to the 25% play speed decrease, health drain occurs 25% slower.
```

- [Difficulty Adjust (DA)](/wiki/Gameplay/Game_modifier/Difficulty_Adjust): Allows manual adjustment of the HP value (range 0–10, or up to 11 with Extended Limits enabled).

## References

[^DifficultyRange]: [`IBeatmapDifficultyInfo` in osu!(lazer)'s source code](https://github.com/ppy/osu/blob/master/osu.Game/Beatmaps/IBeatmapDifficultyInfo.cs)

> [^ManiaHealthProcessor]:  [`ManiaHealthProcessor` in osu!(lazer)'s source code](https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Mania/Scoring/ManiaHealthProcessor.cs)

Reviewer: ```suggestion
[^ManiaHealthProcessor]: [`ManiaHealthProcessor` in osu!(lazer)'s source code](https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Mania/Scoring/ManiaHealthProcessor.cs)
```

---
stub: true
tags:
  - health drain rate
  - passive drain rate
---

> # HP drain rate (Lazer)

Reviewer: ```suggestion
# HP drain rate (lazer)
```

---
stub: true
tags:
  - health drain rate
  - passive drain rate
---

# HP drain rate (lazer)

> <!-- TODO: In-game, this attribute is now called "HP Drain"; this may require updating. -->

Reviewer 4: I'd remove the TODO and use it in the leading section instead, keeping the title unchanged:

```md
**HP Drain** (***HP***) is a [beatmap](/wiki/Beatmap) difficulty setting that controls ...
```

::: alert-note
**Note:** For the osu!(stable) version of this article, see: [HP drain rate](/wiki/Beatmap/HP_drain_rate)
:::

::: alert-note
**Note:** For the gameplay meaning of health, see: [Health](/wiki/Gameplay/Health)
:::

::: alert-note
**Note:** For regulations surrounding HP drain rate, see [Ranking criteria](/wiki/Ranking_criteria)
:::

**HP drain rate** (***HP***) is a [beatmap](/wiki/Beatmap) difficulty setting that controls how much [health](/wiki/Gameplay/Health) is passively lost to health drain. It also affects how heavily a player is penalised for missing notes and how much health is gained back by accurately hitting [hit objects](/wiki/Gameplay/Hit_object).

HP ranges from 0 to 10, (extendable to 11 via the [Difficulty Adjust (DA)](/wiki/Gameplay/Game_modifier/Difficulty_Adjust) mod), where 10 (or 11) is the most punishing and 0 is the most lenient.

> ## Performance in various game modes

Reviewer 4: ```suggestion
## Mechanics
```

Author: Sorry, I forgot this when I was batch adopting suggestions. I'll update and add it manually.

### osu!

<!-- TODO: quantify the relationship between passive HP drain rate and the value -->

In [osu!](/wiki/Game_mode/osu!), the HP value affects both the passive drain rate and the judgement penalties/rewards.

> The health recovery amounts for GREAT, OK and MEH are fixed at +3.00%, +1.10%, and +0.20% respectively, and are not affected by the HP value.

Reviewer 4: ```suggestion
The health recovery amounts for GREAT, OK and MEH are fixed at +3.00%, +1.10%, and +0.20% respectively, and are not affected by the HP value. Slider ticks give +1.5%, while slider reverses and tails give +2%. Slider heads are judged like [hit circles](/wiki/Gameplay/Hit_object/Hit_circle).[^OsuHealthProcessor]
```

with

```md
[^OsuHealthProcessor]: [`OsuHealthProcessor` in osu!(lazer)'s source code](https://github.com/ppy/osu/blob/9f227ed28b6c8ba46dfea1f000f778d8b2827ad0/osu.Game.Rulesets.Osu/Scoring/OsuHealthProcessor.cs)
```

Reviewer 4: I think it's worth mentioning exact recovery percentages here for the full combo, maybe it also resolves the TODO above:

```md
Completing a combo grants extra health after the last object has been hit. Thge bonus is +7% if every object in it was a GREAT (a [Geki](/wiki/Gameplay/Judgement/Geki)), +5% if all of them were at least an OK (a [Katu](/wiki/Gameplay/Judgement/Katu)), and +3% otherwise. Missing a slider tick or dropping a slider tail brings the bonus down to +5%.
```

> The health penalty for MISS is linearly affected by the HP value[^DifficultyRange].

Reviewer 4: ```suggestion
The health penalty for MISS is linearly affected by the HP value:[^DifficultyRange] -3% at HP 0, -12.5% at HP 5, and -20% at HP 10. Missing a slider tick or a reverse costs less: -2%, -7.5%, and -14% respectively. Dropping a slider tail costs no health.[^OsuHealthProcessor]
```

> <!-- TODO: quantify what the value actually means in osu!taiko -->

Reviewer 4: taiko is actually the simplest I think

```suggestion
### osu!taiko

In [osu!taiko](/wiki/Game_mode/osu!taiko), there is no passive health drain. The HP value scales both the health gained from hits and the health lost on misses: the higher it is, the less each hit gives back and the more each miss takes away.[^TaikoHealthProcessor]
```

where

```md
[^TaikoHealthProcessor]: [`TaikoHealthProcessor` in osu!(lazer)'s source code](https://github.com/ppy/osu/blob/9f227ed28b6c8ba46dfea1f000f778d8b2827ad0/osu.Game.Rulesets.Taiko/Scoring/TaikoHealthProcessor.cs)
```

### osu!catch

<!-- TODO: quantify the relationship between passive HP drain rate and the value -->

In [osu!catch](/wiki/Game_mode/osu!catch), the HP value also affects both the passive drain rate and the judgement penalties/rewards.

> In this mode, the relationship between the HP value and the passive drain rate is similar to that in [osu!](/wiki/Game_mode/osu!), but the implementation details differ.

Reviewer 4: this really needs specifics because you pulled them out for osu!

```suggestion
In [osu!catch](/wiki/Game_mode/osu!catch), the HP value affects both the passive drain rate and the health changes from judgements. Missing a fruit or a droplet costs the same as a MISS in osu!: -3% at HP 0, -12.5% at HP 5, and -20% at HP 10. Missing a tiny droplet or a banana costs no health and cannot cause a fail.[^CatchHealthProcessor]
```

whjere

```md
[^CatchHealthProcessor]: [`CatchHealthProcessor` in osu!(lazer)'s source code](https://github.com/ppy/osu/blob/9f227ed28b6c8ba46dfea1f000f778d8b2827ad0/osu.Game.Rulesets.Catch/Scoring/CatchHealthProcessor.cs)
```

### osu!mania

In [osu!mania](/wiki/Game_mode/osu!mania), the HP value only affects the health changes caused by judgement results (there is no passive drain over time). The exact values are shown in the table below:[^ManiaHealthProcessor]

| Judgement (HitResult) | Object type (HitObject) | General formula | HP = 0 | HP = 5 | HP = 10 | HP = 11 |
| :-- | :-- | :-- | :-- | :-- | :-- | :-- |
| **PERFECT** | Any | `(0.0055 - HP * 0.0005) * 100%` | +0.55% | +0.30% | +0.05% | **0.00%** |
| **GREAT** | Any | `(0.0050 - HP * 0.0005) * 100%` | +0.50% | +0.25% | **0.00%** | -0.05% |

> | **GOOD** | Any | `(0.0040 - HP * 0.0004) * 100%` | +0.40% | +0.20% | **0.00%** | -0.04% |

Reviewer 4: what is the purpose of bold formatting here?

| **OK** | Any | `0` | 0.00% | 0.00% | 0.00% | 0.00% |
| **MEH** | Any | `[-(HP + 1) * 0.0016] * 100%` | -0.16% | -0.96% | -1.76% | -1.92% |
| **MISS** | Normal note | `[-(HP + 1) * 0.0075] * 100%` | -0.75% | -4.50% | -8.25% | -9.00% |
| **MISS** | Head or tail of a [hold note](/wiki/Beatmap/Pattern/osu!mania/Hold_note) | `[-(HP + 1) * 0.00375] * 100%` | -0.375% | -2.25% | -4.125% | -4.50% |

## Mod effects

> There are seven mods that alter the HP drain rate when activated:

Reviewer 4: since you added DT/HT, Wind Up, Wind Down, and Adaptive Speed also belong here

Author: There's no entry for Adaptive Speed, and I can't find the corresponding source code to understand its function. My only guess, based on gameplay, is that it slows the player down when they make a poor playthrough and speeds them up when they make a good one.

Reviewer 4: https://osu.ppy.sh/wiki/en/Gameplay/Game_modifier/Adaptive_Speed

- [Easy (EZ)](/wiki/Gameplay/Game_modifier/Easy_(lazer)): Halves the HP value.
- [Hard Rock (HR)](/wiki/Gameplay/Game_modifier/Hard_Rock_(lazer)): Multiplies the HP value by 1.4, up to a maximum of 10.
- [Double Time (DT)](/wiki/Gameplay/Game_modifier/Double_Time_(lazer)) / [Nightcore (NC)](/wiki/Gameplay/Game_modifier/Nightcore_(lazer)): The HP value is not affected, but due to the 50% play speed increase, health drain occurs 50% faster.
- [Half Time (HT)](/wiki/Gameplay/Game_modifier/Half_Time_(lazer)) / [Daycore (DC)](/wiki/Gameplay/Game_modifier/Daycore): The HP value is not affected, but due to the 25% play speed decrease, health drain occurs 25% slower.

> - [Difficulty Adjust (DA)](/wiki/Gameplay/Game_modifier/Difficulty_Adjust): Allows manual adjustment of the HP value (range 0–10, or up to 11 with Extended Limits enabled).

Reviewer 4: ```suggestion
- [Difficulty Adjust (DA)](/wiki/Gameplay/Game_modifier/Difficulty_Adjust): Allows manual adjustment of the HP value (range 0–10 or up to 11 with `Extended Limits` enabled).
```

## References

[^DifficultyRange]: [`IBeatmapDifficultyInfo` in osu!(lazer)'s source code](https://github.com/ppy/osu/blob/master/osu.Game/Beatmaps/IBeatmapDifficultyInfo.cs)

> [^ManiaHealthProcessor]: [`ManiaHealthProcessor` in osu!(lazer)'s source code](https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Mania/Scoring/ManiaHealthProcessor.cs)

Reviewer 4: please add references to specific commits, last option:

<img width="315" height="268" alt="Image" src="https://github.com/user-attachments/assets/5f61f9a7-8dd7-4fdf-9037-82f4085c4bbd" />

```suggestion
[^DifficultyRange]: [`IBeatmapDifficultyInfo` in osu!(lazer)'s source code](https://github.com/ppy/osu/blob/3c1c96f742e7aae2ff67a7361e058fe91ca3b955/osu.Game/Beatmaps/IBeatmapDifficultyInfo.cs)
[^ManiaHealthProcessor]: [`ManiaHealthProcessor` in osu!(lazer)'s source code](https://github.com/ppy/osu/blob/3c1c96f742e7aae2ff67a7361e058fe91ca3b955/osu.Game.Rulesets.Mania/Scoring/ManiaHealthProcessor.cs)
```
